ext/pcntl/pcntl.c

```diff
@@ -385,6 +403,45 @@ PHP_FUNCTION(pcntl_waitpid)
 }
 /* }}} */
 
+#ifdef HAVE_WAITID
+PHP_FUNCTION(pcntl_waitid)
+{
+	zend_long idtype = P_PID;
+	zend_long id = 0;
+	bool id_is_null = 1;
+	zval *user_siginfo = NULL;
+	zend_long options = WEXITED;
+
+	ZEND_PARSE_PARAMETERS_START(2, 4)
+		Z_PARAM_LONG(idtype)
+		Z_PARAM_LONG_OR_NULL(id, id_is_null)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_ZVAL(user_siginfo)
+		Z_PARAM_LONG(options)
+	ZEND_PARSE_PARAMETERS_END();
+
+	errno = 0;
+	siginfo_t siginfo;
+
+	int status = waitid((idtype_t) idtype, (id_t) id, &siginfo, (int) options);
+
+	if (status == -1) {
+		if (errno == EINVAL) {
+			zend_value_error("An invalid value was specified for options, or idtype and id specify an invalid set of processes");
+			RETURN_THROWS();
+		}
+		PCNTL_G(last_error) = errno;
+		php_error_docref(NULL, E_WARNING, "%s", strerror(errno));
+		RETURN_FALSE;
+	}
```

Comment:
The value errors should be triggered *prior* to the call to waitid IMHO.

Moreover, if you pass `null` for the ID and the option expects an actual int ID this should also throw.

Reply:
@Girgias If you pass `null` for the id and the `idtype` requires an actual ID, the function will throw.

Please elaborate why you think value errors should be thrown before the call to `waitid`.

Reply:
it is for value check quality control before doing the call so we do not pass know range of invalid values.

Reply:
If we pass invalid values, [POSIX standard mandates](url) that the system API implementation should fail with EINVAL. Thus, the underlying implementation is mandated to perform the validation for us.

Duplicating the validation code by maintaining a copy of it in the PHP standard library would not be in line with the [Unix software philosophy](https://en.wikipedia.org/wiki/Unix_philosophy#Eric_Raymond's_17_Unix_Rules). Such a design would not follow principles of modularity (or [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns)), composition, Bertrand Meyer's [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), [single source of truth](https://en.wikipedia.org/wiki/Single_source_of_truth) etc.

Reply:
You can have a look at the posix extension, e.g. posix_mknod, there is some pre-validations.

Reply:
I never copy any code or software design without thoroughly reviewing it and understanding what it does and why.

@devnexen  You sent me on an errand to read and review the implementation of `posix_mknod`, and it took me at least an hour (reading the POSIX docs, reading the man pages, checking underlying implementations on different platforms) to do so. As I will elaborate shortly, I found this particular errand to be a waste of time. The first question that comes to mind is: did you thoroughly review the function you referenced (`posix_mknod`) yourself at some point, and found that it is relevant and of exemplary design?

Now to elaborate, by reviewing the code of `posix_mknod`. It does three checks:

1. Checks if the `open_basedir` `php.ini` setting restricts access to the path where user wants to create the file. This check is legitimate, but completely irrelevant to `pcntl_waitid`, as `waitid` has nothing to do with the file system (hence `open_basedir` restrictions are not applicable).

2. If checks if the user is attempting to create a block or character device file with a major number of 0. First thing to notice here is that major and minor device numbers are [not portable per POSIX](https://pubs.opengroup.org/onlinepubs/009696699/functions/mknod.html). But most importantly, I see no reason why PHP should prevent the user from create a device file with major number 0?! For example, in OpenBSD, /dev/console is major 0, minor 0, so at least on one system PHP will prevent the user from performing a perfectly legitimate action! Am I missing something, do you feel that this check is legitimate and relevant?

3. It checks for the presence of the `makedev(3)` utility macro. `makedev` macro implementation(s) could've been duplicated in-line, but the `posix_mknod` developer chose not to do so, instead requiring requiring the macro to be present if the user is creating a S_IFCHR or S_IFBLK file. This example reinforces my argument that you should not duplicate existing system functionality in PHP standard library, and does so very strongly because `makedev` is not standardizes or mandated by POSIX! Yet PHP will refuse to create the dev param itself if libc macro is not present!  `pcntl_waitid` itself of course does not need any aditional functions or macros, so this check is relevant only as an example showing that it is perfectly fine not to duplicate system code.

4. Finally, if mknod fails, it exposes the errno to the user (without converting it to a ValueError, tho). This is legitimate, and this is similar to waitid does.

To summarize, in addition to checking errno, posix_mknod makes one checks that is  completely irrelevant to our discussion (the php.ini restriction feature), one check that avoids adding code to PHP (relying on a utility macro), one check (major == 0) that I would consider problematic: e.g. on OpenBSD it's an outright bug. In short, I see nothing relevant to learn from posix_mknod code that we could use to improve pcntl_waitid code.

Reply:
It was an example, for sure not the same case .. there is [gd](https://github.com/php/php-src/blob/678a481f62db7776156bc69eaa87aaf2ec9e6673/ext/gd/gd.c#L1956) here when we control the quality before the call.

Reply:
Also part of it is in some platforms, a call does not necessarily fully follow the protocol or some implementations are buggy.

Reply:
> there is [gd](https://github.com/php/php-src/blob/678a481f62db7776156bc69eaa87aaf2ec9e6673/ext/gd/gd.c#L1956) here when we control the quality before the call.

And there's a good reason for that -- the gd API is poorly designed in the sense that the export function [gdImagePngEx](https://libgd.github.io/manuals/2.2.2/files/gd_png-c.html#gdImagePngEx) will neither do input validation/sanitization, nor report back an error to the user. It will pass it to libpng, which will in turn pass it to zlib, but not before creating the file and outputting the PNG header. zlib will validate the compression level and fail and user will end up with a broken png file, consisting of only the PNP header (no zlib data chunks as the program failed on first call to zlib).

PHP dev decided to throw an error to the user where the underlying implementation would just produce some stderr messages + a broken file. Absolutely agree with the choice there.

In the gd example, PHP is adding input validation because it validation is not part of the direct underlying API (it happens many stack frames below), so it is much easier and more elegant than having to grep through own stderr and delete failed export files.

In our waitid example, validation is mandated by the underlying API, and validation results are readily available through said API.

It's a very different case.

If you can  point out, very concretely, at particular buggy or non-standard implementations of `waitid(2)`, and how they should be addressed, in as much technical detail as I am offering in my explanations, I'll be happy to make such concrete changes. If not, then let's please stop this licitation of random code examples that are not relevant to the code at hand. They only distract us from finding actual existing issues with the PR.